diff --git a/indoor/indoor/config.py b/indoor/indoor/config.py
index add54e8..7b9236d 100644
--- a/indoor/indoor/config.py
+++ b/indoor/indoor/config.py
@@ -47,7 +47,7 @@ class Config(Blackboard):
         self.camera_down_source: str = 'opencv'
         self.camera_down_topic: str = '/down_camera'
         self.camera_down_is_compressed: bool = True
-        self.camera_down_id: int = 6
+        self.camera_down_id: int = 0
 
         ### Takeoff ###
         self.takeoff_alt: float = 1.2  # meters
@@ -70,7 +70,7 @@ class Config(Blackboard):
 
         # GoToObstacle
         self.obstacle_start_x: float = 0.0  # meters
-        self.obstacle_start_y: float = -3.2  # meters
+        self.obstacle_start_y: float = -3.5  # meters
 
         # Window
         self.obstacle_gate_first_skip: bool = False
@@ -105,8 +105,8 @@ class Config(Blackboard):
         self.inspect_start_time: Time | None = None
 
         # GoToWindow
-        self.inspect_start_x: float = 10.0  # meters
-        self.inspect_start_y: float = 2.0  # meters
+        self.inspect_start_x: float = 9.30  # meters
+        self.inspect_start_y: float = 1.40  # meters
 
         # FindWindow
         self.inspect_back_speed: float = -0.5
diff --git a/indoor/indoor/missions/inspect_sm.py b/indoor/indoor/missions/inspect_sm.py
index 0fa013d..60e75a1 100644
--- a/indoor/indoor/missions/inspect_sm.py
+++ b/indoor/indoor/missions/inspect_sm.py
@@ -18,7 +18,7 @@ class InspectSM(StateMachine):
             'GO_TO_WINDOW',
             GoToWindow(),
             transitions={
-                SUCCEED: 'FIND_WINDOW',
+                SUCCEED: 'COUNT_BABIES',
                 CANCEL: CANCEL,
                 TIMEOUT: CANCEL,
             },
diff --git a/indoor/indoor/states/inspect/count_babies.py b/indoor/indoor/states/inspect/count_babies.py
index 28debd5..ee8f08f 100644
--- a/indoor/indoor/states/inspect/count_babies.py
+++ b/indoor/indoor/states/inspect/count_babies.py
@@ -27,22 +27,23 @@ class CountBabies(State):
         handler: ImageHandler = blackboard.get('image_handler_front')
         handler.image_processing_callback = blackboard.get('callback_baby')
 
-        result: DetectionResult = handler.take_photo()
-
-        if result is not None:
-            babies = result.filter_by_class(
-                [blackboard.get('model_baby_classes_names')])
-            boxes = self._merge_overlapping(
-                babies,
-                blackboard.get('model_baby_overlap_iou'),
-            )
-
-            yasmin.YASMIN_LOG_INFO(f'Number of babies: {len(boxes)}.')
-            for i, box in enumerate(boxes):
-                yasmin.YASMIN_LOG_INFO(f'Baby {i}: bbox={box}')
-
-            blackboard.set('inspect_babies_count', len(boxes))
-            blackboard.set('inspect_babies_boxes', boxes)
+        while True:
+            result: DetectionResult = handler.take_photo()
+
+            if result is not None:
+                babies = result.filter_by_class(
+                    [blackboard.get('model_baby_classes_names')])
+                boxes = self._merge_overlapping(
+                    babies,
+                    blackboard.get('model_baby_overlap_iou'),
+                )
+
+                yasmin.YASMIN_LOG_INFO(f'Number of babies: {len(boxes)}.')
+                for i, box in enumerate(boxes):
+                    yasmin.YASMIN_LOG_INFO(f'Baby {i}: bbox={box}')
+
+                blackboard.set('inspect_babies_count', len(boxes))
+                blackboard.set('inspect_babies_boxes', boxes)
 
         return SUCCEED
 
diff --git a/indoor/indoor/states/inspect/go_to_window.py b/indoor/indoor/states/inspect/go_to_window.py
index 8fd2e0c..973f65a 100644
--- a/indoor/indoor/states/inspect/go_to_window.py
+++ b/indoor/indoor/states/inspect/go_to_window.py
@@ -51,9 +51,8 @@ class GoToWindow(State):
         alt: float = blackboard.get('obstacle_gate_alt')
 
         points = [
-            (None, None, alt),
-            (None, start_y, alt),
-            (start_x, start_y, alt),
+            (start_x, start_y, None),
+            (start_x, start_y, alt)
         ]
 
         if self.check_timeout():
diff --git a/indoor/indoor/states/obstacle/blue_bar.py b/indoor/indoor/states/obstacle/blue_bar.py
index f970e9e..8b1f567 100644
--- a/indoor/indoor/states/obstacle/blue_bar.py
+++ b/indoor/indoor/states/obstacle/blue_bar.py
@@ -5,7 +5,11 @@ from yasmin import State, Blackboard
 from yasmin_ros.yasmin_node import YasminNode
 from yasmin_ros.basic_outcomes import SUCCEED, TIMEOUT
 
-from nectar.control import MavlinkDrone, MoveReference
+from nectar.control import MavlinkDrone, MoveReference, PIDController
+from nectar.vision import LineDetector, ImageHandler
+
+import cv2
+import math
 
 
 class BlueBar(State):
@@ -32,6 +36,12 @@ class BlueBar(State):
         self.add_input_key('obstacle_blue_1')
         self.add_input_key('obstacle_blue_2')
 
+        self.add_input_key('image_handler_down')
+
+        self.add_input_key('obstacle_xy_kp')
+        self.add_input_key('obstacle_xy_kd')
+        self.add_input_key('obstacle_xy_ki')
+
     def execute(self, blackboard: Blackboard):
         drone: MavlinkDrone = blackboard.get('drone')
 
@@ -45,6 +55,8 @@ class BlueBar(State):
         start_x: float = blackboard.get('obstacle_start_x')
         start_y: float = blackboard.get('obstacle_start_y')
 
+        down_handler: ImageHandler = blackboard.get('image_handler_down')
+
         match blackboard.get('obstacle_blue_1'):
             case 1:
                 alt_1 = blackboard.get('obstacle_blue_step_alt_1')
@@ -66,27 +78,78 @@ class BlueBar(State):
                 alt_2 = safe_alt
 
         points = [
-            (start_x+2.25, start_y, alt_1),
-            (start_x+3.25, start_y, alt_1),
-            (start_x+3.25, start_y, alt_2),
-            (start_x+4.25, start_y, alt_2),
+            (0, 0, alt_1),
+            (1.0, 0, alt_1),
+            (0, 0, alt_2),
+            (1.0, 0, alt_2),
         ]
 
         if self.check_timeout():
             return TIMEOUT
 
+        blue_line = LineDetector(
+            color="blue",
+        )
+
+        red_line = LineDetector(
+            color="red"
+        )
+
+        centrilized = False
+        error = -100.0
+        img = down_handler.take_photo()
+
+        pid_x = PIDController(
+            kp=blackboard.get('obstacle_xy_kp'),
+            kd=blackboard.get('obstacle_xy_kd'),
+            ki=blackboard.get('obstacle_xy_ki'),
+            setpoint= img.shape[0]/2 if (img is not None) else 0,
+            output_limits=(-0.3, 0.3)
+        )
+
+        while not centrilized:
+
+            if abs(error) < 20 and error != 0.0:
+                centrilized = True
+                yasmin.YASMIN_LOG_INFO(f"Centrilized between lines | Erro {error}")
+                break
+
+            img = down_handler.take_photo()
+
+            blue_img, blue_mask, blue_cx, blue_cy, blue_angle, blue_w, blue_h = blue_line.detect_line(img)
+            red_img, red_mask, red_cx, red_cy, red_angle, red_w, red_h = red_line.detect_line(img)
+
+            cv2.imwrite(f"/home/jetson/blue_line/f{self.node.get_clock().now()}.jpg", blue_img)
+            cv2.imwrite(f"/home/jetson/red_line/f{self.node.get_clock().now()}.jpg", red_img)            
+
+            error = pid_x._last_error
+
+            if math.isnan(red_cy):
+                red_cy = 0
+
+            yasmin.YASMIN_LOG_INFO(f"Blue center: {blue_cy} | Red center: {red_cy} | Error: {error}")
+
+            if blue_cy:
+                vx = pid_x.update((blue_cy + red_cy)/2)
+            else:
+                yasmin.YASMIN_LOG_INFO("Blue line not detected")
+                vx=0
+                break
+
+            drone.move_velocity(vx=vx)
+
         for x, y, z in points:
-            yasmin.YASMIN_LOG_INFO(f'Fly to x={x}; y={y}; z={z}...')
+            yasmin.YASMIN_LOG_INFO(f"Fly to x = {x} y = {y} z = {z}")
             drone.move_to(
                 x=x,
                 y=y,
+                reference=MoveReference.BODY
+            )
+            drone.move_to(
                 z=z,
-                yaw=0,
-                reference=MoveReference.TAKEOFF,
+                reference=MoveReference.TAKEOFF
             )
 
-            if self.check_timeout():
-                return TIMEOUT
 
         return SUCCEED
 
diff --git a/indoor/indoor/states/obstacle/reacquire_window.py b/indoor/indoor/states/obstacle/reacquire_window.py
index 9ad4734..c808d0a 100644
--- a/indoor/indoor/states/obstacle/reacquire_window.py
+++ b/indoor/indoor/states/obstacle/reacquire_window.py
@@ -39,6 +39,7 @@ class ReacquireWindow(State):
         self.add_input_key('obstacle_gate_alt')
 
     def execute(self, blackboard: Blackboard):
+
         self.timeout: int = blackboard.get('timeout')
         self.mission_timeout: int = blackboard.get('obstacle_timeout')
 
diff --git a/indoor/indoor/states/obstacle/red_bar.py b/indoor/indoor/states/obstacle/red_bar.py
index 48a8810..9fcf370 100644
--- a/indoor/indoor/states/obstacle/red_bar.py
+++ b/indoor/indoor/states/obstacle/red_bar.py
@@ -56,7 +56,7 @@ class RedBar(State):
 
         points = [
             (start_x+1.25, start_y, alt),
-            (start_x+2.25, start_y, alt),
+            (start_x+2.0, start_y, alt),
         ]
 
         if self.check_timeout():
diff --git a/indoor/indoor/states/obstacle/window.py b/indoor/indoor/states/obstacle/window.py
index 33d23e5..4e785ad 100644
--- a/indoor/indoor/states/obstacle/window.py
+++ b/indoor/indoor/states/obstacle/window.py
@@ -129,7 +129,7 @@ class Window(State):
                     f'| Error: y={error_y:.0f}; z={error_z:.0f}. '
                     f'| Output: y={output_y:.1f}; z={output_z:.1f}. '
                 )
-                drone.move_velocity(vy=output_y, vz=output_z)
+                drone.move_velocity(0.1, vy=output_y, vz=output_z)
 
             if aligned >= aligned_threshold:
                 break
@@ -162,13 +162,22 @@ class Window(State):
             return TIMEOUT
 
         yasmin.YASMIN_LOG_INFO('Fly through the window.')
-        drone.move_to(
-            x=1.0,
-            y=None,
-            z=0,
-            yaw=0,
-            reference=MoveReference.TAKEOFF
-        )
+
+        if self.position == 'first':
+            drone.move_to(
+                x=0.75,
+                y=0.0,
+                z=0,
+                yaw=0,
+                reference=MoveReference.BODY
+            )
+        else:
+            drone.move_to(
+                x=6.25,
+                y=0.0,
+                z=0,
+                reference=MoveReference.TAKEOFF
+            )
 
         return SUCCEED
 
